service/matching/taskQueueManager.go

```diff
@@ -479,14 +482,31 @@ func (c *taskQueueManagerImpl) GetUserData(ctx context.Context) (*persistencespb
 	return data, err
 }
 
-func (c *taskQueueManagerImpl) UpdateUserData(ctx context.Context, updateFn func(*persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, error)) error {
+//nolint:revive // control coupling
+func (c *taskQueueManagerImpl) UpdateUserData(ctx context.Context, replicate bool, updateFn func(*persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, error)) error {
 	newData, err := c.db.UpdateUserData(ctx, updateFn)
 	c.signalIfFatal(err)
 	if err != nil {
 		return err
 	}
+	if replicate && c.namespaceReplicationQueue != nil {
+		err = c.namespaceReplicationQueue.Publish(ctx, &repication.ReplicationTask{
+			TaskType: enumsspb.REPLICATION_TASK_TYPE_TASK_QUEUE_USER_DATA,
+			Attributes: &repication.ReplicationTask_TaskQueueUserDataAttributes{
+				TaskQueueUserDataAttributes: &repication.TaskQueueUserDataAttributes{
+					NamespaceId:   c.taskQueueID.namespaceID.String(),
+					TaskQueueName: c.taskQueueID.BaseNameString(),
+					UserData:      newData.GetData(),
+				},
+			},
+		})
+		if err != nil {
+			c.logger.Error("Failed to publish a replication task after updating task queue user data", tag.Error(err))
+			return serviceerror.NewUnavailable("storing task queue user data succeeded but publishing to the namespace replication queue failed, please try again")
```

Comment:
Try what again? This message gets bubbled all the way back to the user making a change to ID sets?

Is the idea to get the replication task to get queued properly by having them do the same operation again? If so, that seems less than ideal - because from their perspective the operation "succeeded" in the sense that whatever they did is now reflected on the cluster they are working with.

Is there some other place we can spool this to so it gets retried? I'm assuming probably not, in which case I don't know that I have any better idea here but just wanted to ask

Reply:
This is what I want to do for the first round, after that I'll wrap this operation in a workflow that returns once applying the change to the task_queue_user_data table succeeds and continues to enqueue the replication event asynchronously.